### PR TITLE
Adds rewrite phase to aggregations

### LINKED ISSUE #17676: Add rewrite phase to Aggregations
It would be useful to have a rewrite phase on aggregations so we can:
1. optimise an aggregation to make it run better (e.g. rewrite the `QueryBuilder`s in the filter/filters aggregation to enable better request caching)
2. Add syntactic sugar around aggregations (see https://github.com/elastic/elasticsearch/issues/17631 for an example)

We could add a rewrite method to `AggregationBuilder` which would be called from the `SearchSourceBuilder#rewrite` method.

### PR DESCRIPTION
**Pull request by colings86**
This change adds aggregations to the rewrite performed by the `SearchSourceBuilder`. This means that `AggregationBuilder`s are able to implement a `rewrite()` method where they can return a new `AggregationBuilder` which is functionally the same but in a more primitive form. This is exactly analogous to the rewrite done by the `QueryBuilder`s.

The first aggregation to implement the rewrite are the filter and filters aggregations so they can rewrite the filters they contain.

Closes #17676

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java [hunk 1]
```diff
@@ -95,6 +96,23 @@ public String getName() {
...
+        }
+        if (getMetaData() != null && rewritten.getMetaData() == null) {
+            rewritten.setMetaData(getMetaData());
+        }
+        AggregatorFactories.Builder rewrittenSubAggs = factoriesBuilder.rewrite(context);
+        rewritten.subAggregations(rewrittenSubAggs);
+        return rewritten;
+    }
+
+    protected AggregationBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java]: Javadoc would be nice I think.

File: core/src/test/java/org/elasticsearch/indices/IndicesRequestCacheIT.java [hunk 2]
```diff
@@ -411,18 +410,7 @@ public void testCanCache() throws Exception {
...
-                .addAggregation(filter("foo", QueryBuilders.rangeQuery("s").from("now-10y").to("now"))).get();
-        assertSearchResponse(r4);
-        assertThat(r4.getHits().getTotalHits(), equalTo(7L));
-        assertThat(client().admin().indices().prepareStats("index").setRequestCache(true).get().getTotal().getRequestCache().getHitCount(),
-                equalTo(0L));
-        assertThat(client().admin().indices().prepareStats("index").setRequestCache(true).get().getTotal().getRequestCache().getMissCount(),
-                equalTo(0L));
-
-        // If the request has an aggregation containng now we should not cache
+        // If the request has an non-filter aggregation containng now we should not cache
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/IndicesRequestCacheIT.java]: s/containng/containing/

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/50_filter.yml [hunk 3]
```diff
@@ -32,7 +32,8 @@ setup:
       indices.refresh: {}
 
 ---
-"Filter aggs with terms lookup ensure not cached":
+"Filter aggs with terms lookup and ensure it's cached":
+  # Because the filter agg rewrites the terms lookup in the rewrite phase the request can be cached
   - skip:
       version: " - 5.0.0"
```
Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/50_filter.yml]: I think you have to skip before 6.0 because a mixed cluster may or may not cache depending on where the shards live.

**Review by nik9000 — COMMENTED** [review]
Neat!

File: core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java [hunk 4]
```diff
@@ -118,7 +118,7 @@ public AB setMetaData(Map<String, Object> metaData) {
 
     @Override
     public Map<String, Object> getMetaData() {
-        return Collections.unmodifiableMap(metaData);
+        return metaData == null ? null : Collections.unmodifiableMap(metaData);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java]: should we rather return an empty map here?

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java [hunk 5]
```diff
@@ -95,6 +96,23 @@ public String getName() {
     @Override
     public abstract AggregationBuilder subAggregations(AggregatorFactories.Builder subFactories);
 
+    public final AggregationBuilder rewrite(QueryRewriteContext context) throws IOException {
+        AggregationBuilder rewritten = doRewrite(context);
+        if (rewritten == this) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java]: hmm what if the rewritten agg needs to be rewritten? I think we need to do this in a loop as we did with the queries?

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java [hunk 6]
```diff
@@ -466,5 +467,32 @@ public boolean equals(Object obj) {
                 return false;
             return true;
         }
+
+        /**
+         * Rewrites the underlying aggregation builders into their primitive
+         * form. If the builder did not change the identity reference must be
+         * returned otherwise the builder will be rewritten infinitely.
+         */
+        public Builder rewrite(QueryRewriteContext context) throws IOException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java]: can we get a unittest for this?

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java (hunk 5 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java]: Agreed, though I put the loop in AggregatorFactories.Builder so that it loops over completely rewritten agg builders.

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java (hunk 6 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java]: Added

**Comment by colings86** [comment]
@s1monw @nik9000 thanks for the reviews, I added a commit that should address your comments, could you take another look?

File: core/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesTests.java [hunk 7]
```diff
@@ -244,6 +254,44 @@ public void testMissingType() throws Exception {
...
+        FilterAggregationBuilder filterAggBuilder = new FilterAggregationBuilder("titles", new WrapperQueryBuilder(bytesReference));
+        BucketScriptPipelineAggregationBuilder pipelineAgg = new BucketScriptPipelineAggregationBuilder("const", new Script("1"));
+        AggregatorFactories.Builder builder = new AggregatorFactories.Builder().addAggregator(filterAggBuilder)
+                .addPipelineAggregator(pipelineAgg);
+        AggregatorFactories.Builder rewritten = builder
+                .rewrite(new QueryRewriteContext(null, null, null, xContentRegistry, null, null, () -> 0L));
+        assertNotSame(builder, rewritten);
+        List<AggregationBuilder> aggregatorFactories = rewritten.getAggregatorFactories();
+        assertEquals(1, aggregatorFactories.size());
+        assertThat(aggregatorFactories.get(0), Matchers.instanceOf(FilterAggregationBuilder.class));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesTests.java]: I think it is more idiomatic to statically import the `instanceOf`.

File: core/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesTests.java [hunk 8]
```diff
@@ -244,6 +254,44 @@ public void testMissingType() throws Exception {
...
+        // Check the filter was rewritten from a wrapper query to a terms query
+        QueryBuilder rewrittenFilter = rewrittenFilterAggBuilder.getFilter();
+        assertThat(rewrittenFilter, Matchers.instanceOf(TermsQueryBuilder.class));
+        
+        // Check that a further rewrite returns the same aggregation factories builder
+        AggregatorFactories.Builder secondRewritten = rewritten
+                .rewrite(new QueryRewriteContext(null, null, null, xContentRegistry, null, null, () -> 0L));
+        assertSame(rewritten, secondRewritten);
+    }
+
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesTests.java]: If you want to test the multi-write behavior you could make a testing aggregation here that needs to be rewritten twice. I'm not sure how important that is to you, but it ought to be possible.

**Review by nik9000 — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java [hunk 9]
```diff
@@ -118,7 +118,7 @@ public AB setMetaData(Map<String, Object> metaData) {
 
     @Override
     public Map<String, Object> getMetaData() {
-        return Collections.unmodifiableMap(metaData);
+        return metaData == null ? Collections.emptyMap() : Collections.unmodifiableMap(metaData);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java]: why are we adding leniency?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java]: or is it fixing a bug?

**Review by jpountz — APPROVED** [review]
LGTM but I'm curious as to why we need to null check the metadata. Ideally It would never be null.

File: core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java (hunk 9 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java]: its fixing a bug as the metadata was always null if not set, we just never called `getMetadata()` from within Elasticsearch with an unset metadata so we didn't see the bug. I agree that ideally the metadata would never be null but I'm not sure on the side effects of that change so I think it would be better to explore that in a different PR. Does that work for you?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/AbstractAggregationBuilder.java]: sure
